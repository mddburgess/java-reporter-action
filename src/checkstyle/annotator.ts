import CheckstyleReport, {CheckstyleViolation} from './report';
import {Annotation, AnnotationLevel} from '../common/github';
import path from 'path';

class CheckstyleAnnotator {

    public annotate(report: CheckstyleReport): Annotation[] {
        return report.violations.map(violation => this.annotateViolation(violation));
    }

    private annotateViolation(violation: CheckstyleViolation) {
        return {
            path: this.resolvePath(violation.filePath),
            start_line: violation.line,
            end_line: violation.line,
            start_column: violation.column,
            end_column: violation.column,
            annotation_level: this.resolveAnnotationLevel(violation),
            title: this.resolveTitle(violation),
            message: violation.message
        }
    }

    private resolvePath(filePath: string): string {
        return process.env.GITHUB_WORKSPACE
            ? path.relative(process.env.GITHUB_WORKSPACE, filePath)
            : filePath;
    }

    private resolveAnnotationLevel(violation: CheckstyleViolation): AnnotationLevel {
        switch (violation.severity) {
            case 'error':
                return 'failure';
            case 'warning':
                return 'warning';
            case 'info':
                return 'notice';
        }
    }

    private resolveTitle(violation: CheckstyleViolation): string {
        return violation.rule.split('.').slice(-1)[0];
    }
}

export default CheckstyleAnnotator;
