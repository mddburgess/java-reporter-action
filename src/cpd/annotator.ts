import CpdReport, {CpdDuplication, CpdFile} from './report';
import {Annotation} from '../common/github';
import path from 'path';

class CpdAnnotator {

    public annotate(report: CpdReport): Annotation[] {
        return report.duplications
            .map(duplication => this.annotateDuplication(duplication))
            .reduce((a, b) => a.concat(b));
    }

    private annotateDuplication(duplication: CpdDuplication): Annotation[] {
        duplication.files.forEach(file => file.path = this.resolvePath(file.path));
        return duplication.files.map(file => this.annotateFile(duplication, file));
    }

    private resolvePath(filePath: string): string {
        return process.env.GITHUB_WORKSPACE
            ? path.relative(process.env.GITHUB_WORKSPACE, filePath)
            : filePath;
    }

    private annotateFile(duplication: CpdDuplication, file: CpdFile): Annotation {
        return {
            path: file.path,
            start_line: file.startLine,
            end_line: file.endLine,
            annotation_level: 'warning',
            title: this.resolveTitle(duplication.lines),
            message: this.resolveMessage(duplication, file.path)
        }
    }

    private resolveTitle(lines: number): string {
        return `${lines} duplicated lines`;
    }

    private resolveMessage(duplication: CpdDuplication, path: string): string {
        return [
            `Found ${duplication.lines} lines duplicated at:`,
            ...duplication.files
                .filter(file => file.path !== path)
                .map(file => `\t${file.path} line ${file.startLine}`)
        ].join('\n');
    }
}

export default CpdAnnotator;
