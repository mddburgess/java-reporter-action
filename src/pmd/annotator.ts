import {PmdViolation} from './report';
import {Annotation} from '../common/github';
import path from 'path';

export function toAnnotation(violation: PmdViolation): Annotation {
    return {
        path: resolvePath(violation.filePath),
        start_line: violation.startLine,
        end_line: violation.endLine,
        start_column: violation.startLine === violation.endLine ? violation.startColumn : undefined,
        end_column: violation.startLine === violation.endLine ? violation.endColumn : undefined,
        annotation_level: 'failure',
        title: resolveTitle(violation),
        message: violation.message
    }
}

function resolvePath(filePath: string): string {
    return process.env.GITHUB_WORKSPACE
        ? path.relative(process.env.GITHUB_WORKSPACE, filePath)
        : filePath;
}

function resolveTitle(violation: PmdViolation): string {
    return `${violation.ruleset}: ${violation.rule}`;
}
