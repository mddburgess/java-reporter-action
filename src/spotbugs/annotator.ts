import {Annotation, AnnotationLevel} from '../common/github';
import SpotbugsReport, {SpotbugsBug} from './report';
import {findRelativePath} from '../common/files';

class SpotbugsAnnotator {

    private filePaths = new Map<string, string>();

    public async annotate(report: SpotbugsReport): Promise<Annotation[]> {
        const annotations = [];
        for (const bug of report.bugs) {
            annotations.push(await this.annotateViolation(bug, report.categories));
        }
        return annotations;
    }

    private async annotateViolation(bug: SpotbugsBug, categories: Map<string, string>): Promise<Annotation> {
        return {
            path: this.filePaths.get(bug.filePath) || await this.resolvePath(bug.filePath),
            start_line: bug.startLine,
            end_line: bug.endLine,
            annotation_level: this.resolveAnnotationLevel(bug),
            title: this.resolveTitle(bug, categories),
            message: bug.longMessage
        }
    }

    private async resolvePath(filePath: string) {
        const searchPath = `**/${filePath}`;
        const foundFilePath = await findRelativePath(searchPath);
        this.filePaths.set(filePath, foundFilePath);
        return foundFilePath;
    }

    private resolveAnnotationLevel(bug: SpotbugsBug): AnnotationLevel {
        switch (bug.priority) {
            case 1:
                return 'failure';
            case 2:
                return 'warning';
            default:
                return 'notice';
        }
    }

    private resolveTitle(bug: SpotbugsBug, categories: Map<string, string>): string {
        const category = categories.get(bug.category) || bug.category;
        return `${category}: ${bug.shortMessage}`;
    }
}

export default SpotbugsAnnotator;
