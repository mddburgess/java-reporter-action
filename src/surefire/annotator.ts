import SurefireReport, {SurefireTestCase} from './report';
import {Annotation, AnnotationLevel} from '../common/github';
import {findRelativePath} from '../common/files';

class SurefireAnnotator {

    public async annotate(report: SurefireReport): Promise<Annotation[]> {
        return Promise.all(report.testCases.map(testCase => this.annotateTestCase(testCase)));
    }

    private async annotateTestCase(testCase: SurefireTestCase): Promise<Annotation> {
        const line = this.resolveLine(testCase);
        return {
            path: await this.resolvePath(testCase.className),
            start_line: line,
            end_line: line,
            annotation_level: this.resolveAnnotationLevel(testCase),
            title: this.resolveTitle(testCase),
            message: this.resolveMessage(testCase),
            raw_details: this.resolveRawDetails(testCase)
        };
    }

    private resolvePath(className: string) {
        const searchPath = '**/' + className.split('.').join('/') + '.java';
        return findRelativePath(searchPath);
    }

    private resolveLine(testCase: SurefireTestCase): number {
        if (testCase.stackTrace) {
            const stackTrace = testCase.stackTrace;
            const stackFrames = stackTrace.match(RegExp(`${testCase.className}.*:\\d+`));
            if (stackFrames) {
                const [stackFrame] = stackFrames.slice(-1);
                const [, line] = stackFrame.split(':');
                return Number(line);
            }
        }
        return 0;
    }

    private resolveAnnotationLevel(testCase: SurefireTestCase): AnnotationLevel {
        switch (testCase.result) {
            case 'failure':
            case 'error':
                return 'failure';
            case 'skipped':
                return 'notice';
            case 'success':
            case undefined:
                throw Error('unexpected test case');
        }
    }

    private resolveTitle(testCase: SurefireTestCase): string {
        const [simpleClassName] = testCase.className.split('.').slice(-1);
        return `Test ${testCase.result}: ${simpleClassName}.${testCase.testName}`;
    }

    private resolveMessage(testCase: SurefireTestCase): string {
        return testCase.message || testCase.stackTrace || `Test ${testCase.result}`;
    }

    private resolveRawDetails(testCase: SurefireTestCase): string | undefined {
        return testCase.stackTrace;
    }
}

export default SurefireAnnotator;
