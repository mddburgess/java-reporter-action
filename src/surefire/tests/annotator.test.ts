import {resolveLine, resolvePath, toAnnotation} from '../annotator';

describe('toAnnotation()', () => {

    it('can convert a Surefire test case to a GitHub annotation', async () => {
        const annotation = await toAnnotation({
            className: 'org.example.SimpleTest',
            testName: 'failingTestWithoutMessage',
            result: 'failure',
            message: undefined,
            stackTrace: 'org.opentest4j.AssertionFailedError: \n' +
                '\tat org.example.SimpleTest.failingTestWithoutMessage(SimpleTest.java:24)\n',
        });
        expect(annotation).toEqual({
            path: 'example/src/test/java/org/example/SimpleTest.java',
            start_line: 24,
            end_line: 24,
            annotation_level: 'failure',
            title: 'SimpleTest.failingTestWithoutMessage',
            message: 'org.opentest4j.AssertionFailedError: \n' +
                '\tat org.example.SimpleTest.failingTestWithoutMessage(SimpleTest.java:24)\n',
            raw_details: 'org.opentest4j.AssertionFailedError: \n' +
                '\tat org.example.SimpleTest.failingTestWithoutMessage(SimpleTest.java:24)\n'
        })
    })
})

describe('resolvePath()', () => {

    it('can resolve a path from a class name', async () => {
        const path = await resolvePath('org.example.SimpleTest');
        expect(path).toEqual('example/src/test/java/org/example/SimpleTest.java');
    });
});

describe('resolveLine()', () => {

    it('can resolve a line from a stacktrace', () => {
        const line = resolveLine({
            className: 'org.example.SimpleTest',
            testName: 'failingTestWithoutMessage',
            stackTrace: 'org.opentest4j.AssertionFailedError: \n' +
                '\tat org.example.SimpleTest.failingTestWithoutMessage(SimpleTest.java:24)'
        });
        expect(line).toEqual(24);
    });

    it('returns 0 when a stacktrace is not available', () => {
        const line = resolveLine({
            className: 'org.example.SimpleTest',
            testName: 'disabledTest',
            message: 'void org.example.SimpleTest.disabledTest() is @Disabled'
        });
        expect(line).toEqual(0);
    })
})
