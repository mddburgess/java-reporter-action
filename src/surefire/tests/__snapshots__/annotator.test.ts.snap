// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Surefire annotator can annotate a Surefire report 1`] = `
Array [
  Object {
    "annotation_level": "notice",
    "end_line": 0,
    "message": "void org.example.SimpleTest.disabledTest() is @Disabled",
    "path": "example/src/test/java/org/example/SimpleTest.java",
    "raw_details": undefined,
    "start_line": 0,
    "title": "Test skipped: SimpleTest.disabledTest",
  },
  Object {
    "annotation_level": "failure",
    "end_line": 29,
    "message": "java.lang.RuntimeException: error message
	at org.example.SimpleTest.errorTestWithMessage(SimpleTest.java:29)",
    "path": "example/src/test/java/org/example/SimpleTest.java",
    "raw_details": "java.lang.RuntimeException: error message
	at org.example.SimpleTest.errorTestWithMessage(SimpleTest.java:29)",
    "start_line": 29,
    "title": "Test error: SimpleTest.errorTestWithMessage",
  },
  Object {
    "annotation_level": "failure",
    "end_line": 24,
    "message": "org.opentest4j.AssertionFailedError: 
	at org.example.SimpleTest.failingTestWithoutMessage(SimpleTest.java:24)",
    "path": "example/src/test/java/org/example/SimpleTest.java",
    "raw_details": "org.opentest4j.AssertionFailedError: 
	at org.example.SimpleTest.failingTestWithoutMessage(SimpleTest.java:24)",
    "start_line": 24,
    "title": "Test failure: SimpleTest.failingTestWithoutMessage",
  },
  Object {
    "annotation_level": "notice",
    "end_line": 45,
    "message": "org.opentest4j.TestAbortedException: Assumption failed: assumption is not true
	at org.example.SimpleTest.skippedTest(SimpleTest.java:45)",
    "path": "example/src/test/java/org/example/SimpleTest.java",
    "raw_details": "org.opentest4j.TestAbortedException: Assumption failed: assumption is not true
	at org.example.SimpleTest.skippedTest(SimpleTest.java:45)",
    "start_line": 45,
    "title": "Test skipped: SimpleTest.skippedTest",
  },
  Object {
    "annotation_level": "failure",
    "end_line": 34,
    "message": "java.lang.RuntimeException
	at org.example.SimpleTest.errorTestWithoutMessage(SimpleTest.java:34)",
    "path": "example/src/test/java/org/example/SimpleTest.java",
    "raw_details": "java.lang.RuntimeException
	at org.example.SimpleTest.errorTestWithoutMessage(SimpleTest.java:34)",
    "start_line": 34,
    "title": "Test error: SimpleTest.errorTestWithoutMessage",
  },
  Object {
    "annotation_level": "failure",
    "end_line": 19,
    "message": "org.opentest4j.AssertionFailedError: expected: <5> but was: <4>
	at org.example.SimpleTest.failingTestWithMessage(SimpleTest.java:19)",
    "path": "example/src/test/java/org/example/SimpleTest.java",
    "raw_details": "org.opentest4j.AssertionFailedError: expected: <5> but was: <4>
	at org.example.SimpleTest.failingTestWithMessage(SimpleTest.java:19)",
    "start_line": 19,
    "title": "Test failure: SimpleTest.failingTestWithMessage",
  },
]
`;
